Add topic consumer autocreation

nerhneiro opened the pull request:
### Changelog entry <!-- a user-readable short description of the changes that goes to CHANGELOG.md and Release Notes -->

Add topic and consumer autocreation when auto_create_topics_enable flag is enabled in config.

### Changelog category <!-- remove all except one -->

* Not for changelog (changelog entry is not required)

### Description for reviewers <!-- (optional) description for those who read this PR -->

...

Commit:
generate TListGroupRequestData and TListGroupResponseData classes, EApiKey::LIST_GROUPS

Commit:
add actor files, generate TListGroupsResponseData and TListGroupsRequestData in kaffka_messages.h
write test to check is ListGroupsActor works correctly

Commit:
add statesFilter realization
add new test

Commit:
Merge branch 'add-list-groups-for-akhq' into add-describe-consumer-group-for-akhq

Commit:
generate DESCRIBE_GROUPS in kafka_messages

Commit:
add correct DatabasePath from context
add check on EnableKafkaNativeBalancing feature flag existance
add empty check in up_protocol.cpp
correct codestyle
delete excesive variables
add comments

Commit:
add ApiKey for LIST_GROUPS
change error check from NOT_FOUND to SCHEME_ERROR

Commit:
Merge branch 'add-list-groups-for-akhq' into add-describe-consumer-group-for-akhq

Commit:
add MemberTimeoutsMs struct for deadlines/timeouts
change heartbeat_deadline from now + rebalaceTimeout + sessionTimeout to just now + sessionTimeout
add checks to stop waiting ungracefully finished consumers after sessionTimeout has passed

Commit:
remove surplus space

Commit:
add DescribeGroups actor
shift numbersToStatesMapping and statesToNumbersMapping into a separate file
add test for DescribeGroups in ut_protocol.cpp

Commit:
solve mistakes

Commit:
Merge branch 'fix-bug-leave-group-long-timeout' into add-offset-topic-akhq

Commit:
add correct dying
change order of methods
change ApiKeys.size() check from 18 to 20

Commit:
Merge branch 'main' into add-describe-consumer-group-for-akhq

Commit:
change apiKeys count in BalanceScenarioCdc test

Commit:
Merge branch 'main' into add-describe-consumer-group-for-akhq

Commit:
add topics request for groups in offset_fetch_actor

Commit:
Merge branch 'add-describe-consumer-group-for-akhq' into add-offset-topic-akhq

Commit:
correct byte storage in MemberAssignment
add new field MemberAssignmentStr in TDescribedGroup::TDescribedGroupMember in generator

Commit:
Merge branch 'main' into add-describe-consumer-group-for-akhq

Commit:
correct EXPECTED_API_KEYS_COUNT

Commit:
add correct work for timestamp=0
add correct version for LIST_OFFSETS

Commit:
Merge branch 'add-describe-consumer-group-for-akhq' into add-offset-topic-akhq

Commit:
correct count for EXPECTED_API_KEYS_COUNT

Commit:
add comment for the logic

Commit:
unite several kqp requests into one in describe groups
add check that tables were inited
add test for inited tables in DiscribeGroupsScenario and ListGroupsScenario
reorder methods in list groups
add new method options in kafka_test_client

Commit:
Merge branch 'main' into add-describe-consumer-group-for-akhq

Commit:
Merge branch 'add-describe-consumer-group-for-akhq' into add-offset-topic-akhq

Commit:
unite several kqp requests into one in describe groups
add check that tables were inited
add test for inited tables in DiscribeGroupsScenario and ListGroupsScenario
reorder methods in list groups
add new method options in kafka_test_client

Commit:
reduce copying in kafka_offset_fetch_actor
add ConfigSource return in configEntry

Commit:
add correct name for enum

Commit:
better naming
delete surplus parameters and variables

Commit:
Merge branch 'main' into add-describe-consumer-group-for-akhq

Commit:
Merge branch 'add-describe-consumer-group-for-akhq' of github.com:nerhneiro/ydb into add-describe-consumer-group-for-akhq

Commit:
remove surplus line

Commit:
Merge branch 'main' into add-describe-consumer-group-for-akhq

Commit:
add MemberMetadataStr to avoid memory error

Commit:
Merge branch 'main' into add-offset-topic-akhq

Commit:
Merge branch 'add-describe-consumer-group-for-akhq' into add-offset-topic-akhq

Commit:
unite multiple KQP requests into one

Commit:
Merge branch 'main' into add-offset-topic-akhq

Commit:
remove duplicated lines

Commit:
add logprefix method

Commit:
remove extra prefix in logging

Commit:
add autocreation of topic and consumers
no flag and autorisation checks

Commit:
start writing a new test
add requestLimitNumber in fetch_offsets

Commit:
add AutoCreateTopicsEnable flag
add topic existance check in metadata_actor and offset_fetch_actor
add consumer auto-addition in offset_fetch_actor

Commit:
Merge branch 'main' into add-topic-consumer-autocreation

Commit:
remove duplicate method

Commit:
correct indents

Commit:
add checks on offset fetch only if SetAutoCreateTopicsEnable is false

Commit:
OffsetCommitAndFetch works for nonExistent topic

Commit:
resolve error in OffsetCommitAndFetchScenario test

Commit:
better indents
remove extra lines

Commit:
remove surplus dash

a-serebryanskiy commented on ydb/core/kafka_proxy/ut/ut_protocol.cpp:
```diff
@@ -2516,6 +2563,69 @@ Y_UNIT_TEST_SUITE(KafkaProtocol) {
+        TString nonExistedTopicName2 = "non-existent-topic-2";
+        TString newConsumer = "new-consumer";
+
+        if (testServer.KikimrServer.get()->ServerSettings->AppConfig->GetKafkaProxyConfig().GetAutoCreateTopicsEnable()) {
```
В целом про условия в тесте: это или значительная ошибка или я что-то не понял. Как ты планировала, что альтернативные ветки в тесте будут выполняться? Если я все правильно понял, они просто будут лежать мертвым кодом. А мертвый код в PR-е - это прямо плохо. Такое в будущем допускать нельзя.

Commit:
create separate methods for topic and consumer creation
remove ContextForTopicCreation from class fields
start reformating a test

Commit:
add test for metadata
move topic and consumer creation to separate methods
modify error check
add structs

Commit:
Merge branch 'main' into add-topic-consumer-autocreation

Commit:
Merge branch 'main' into add-topic-consumer-autocreation

Commit:
resolve merge conflict

Commit:
remove excess lines
change russian comments to english

Commit:
remove a surplus line

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_metadata_actor.cpp:
```diff
@@ -248,15 +250,56 @@ void TKafkaMetadataActor::HandleLocationResponse(TEvLocationResponse::TPtr ev, c
+        TActorId child = SendTopicRequest(topicName);
+        TopicIndexes[child].push_back(topicIndex);
+    } else if (InflyCreateTopics == 0) {
+        RespondIfRequired(ctx);
```
Есть тесты на такой случай? Когда какой-то топик не создастся, например, из-за отсутствующих прав?

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_metadata_actor.h:
```diff
@@ -12,6 +12,11 @@
 
 namespace NKafka {
 
+    struct TopicNameToIndex {
```
TTopicNameToIndex - T префикс у всех классов/структур.

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_offset_fetch_actor.cpp:
```diff
@@ -335,45 +432,96 @@ void TKafkaOffsetFetchActor::ParseGroupsAssignments(NKqp::TEvKqp::TEvQueryRespon
+
+    auto message = std::make_shared<NKafka::TCreateTopicsRequestData>();
+    TCreateTopicsRequestData::TCreatableTopic topicToCreate;
+    topicToCreate.Name = topicName;
```
Забыли тут groupId передать видимо.

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_offset_fetch_actor.cpp:
```diff
@@ -233,16 +236,100 @@ void TKafkaOffsetFetchActor::Handle(TEvKafka::TEvCommitedOffsetsResponse::TPtr&
+                break;
+            } else if (topicPartition.ErrorCode == EKafkaErrors::UNKNOWN_TOPIC_OR_PARTITION) {
+                // topic does not exist case
+                bool continueChecking = CreateTopicIfNecessary(topicName, *topicGroupRequest.TopicRequest.Name, topicGroupRequest.GroupId, ctx);
```
Эта ошибка может быть при указании слишком большой партиции, в таком случае пытаться создать новый топик не надо.

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_offset_fetch_actor.h:
```diff
@@ -23,9 +35,32 @@ struct TopicEntities {
     std::shared_ptr<TSet<ui32>> Partitions = std::make_shared<TSet<ui32>>();
 };
 
+struct ModifiedTopicInfo {
```
TModifiedTopicInfo

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_offset_fetch_actor.h:
```diff
@@ -23,9 +35,32 @@ struct TopicEntities {
 
+struct ModifiedTopicInfo {
+    TString TopicName;
+    TopicEntities Entities;
```
TTopicEntities. И остальные классы ниже так же поменять надо

alexnick88 commented on ydb/core/kafka_proxy/actors/kafka_offset_fetch_actor.cpp:
```diff
@@ -233,16 +236,100 @@ void TKafkaOffsetFetchActor::Handle(TEvKafka::TEvCommitedOffsetsResponse::TPtr&
+        return;
+    }
+
+    const ModifiedTopicInfo& alteredTopicInfo = AlterTopicInfo[ev->Cookie];
```
А что если после альтера консумер будет, а партиции -- нет? Хочу тест, когда fetch вызывается по существующему топику без консумера по одной несущетвующей партиции(слишком большой). Вообще решение тут будет сначала проверять наличие топика и партиции, а только потом консумера в дескрайбе топика

Commit:
return error in metadata actor correctly

Commit:
change name for default partitions num in config
add tests on topic autocreation with too big partition

Commit:
add a new test

Commit:
add enableAutoConsumerCreation feature flag
add test on consumer autocreation
remove surplus struct and location requests

Commit:
turn flag AutoCreateConsumersEnable on by default

Commit:
make AutoCreateTopicsEnable true by default

Commit:
resolve mistakes with flags in tests
